Game.management

matthewphilyaw opened the pull request:
Working on the character and game management. Covers #21, #16 , #22
- [x] game model
- [x] character model
- [x] test character and game models
- [x] switch to character_id as the key in the player registry and this includes changing the player topic to be character id
- [x] game controller that supports GET for now
- [x] tests for game controller
- [x] character controller to support CRUD
- [ ] tests for character

matthewphilyaw committed:
initial pass at schemas for game/character.

matthewphilyaw committed:
remove has many relationship out of game and for now out of user

matthewphilyaw committed:
remove tests for game and character for the moment.

matthewphilyaw committed:
added unique constraints to character for so that only one character name can exist per game and added one for game name as well.

matthewphilyaw commented on priv/repo/migrations/20160203035956_create_character.exs:
```diff
@@ -0,0 +1,17 @@
+    end
+    create index(:characters, [:user_id])
+    create index(:characters, [:game_id])
+    create unique_index(:characters, [:name, :game_id])
```
@copenhas we may want to include user_id if we wanted to limit a user to having only one character per game. Thoughts? I'm not too concerned with it at the moment, I figured at least those two fields should be good enough.

copenhas replied:
A user should be allowed to have multiple characters in a game.

matthewphilyaw replied:
Cool, then this constraint satisfies this. It only fires if the same character name is used twice in a game but a user can have multiple characters. I have tests proving all that out (I think I covered the case to ensure one user and two characters are possible as long as the name is different...)

matthewphilyaw commented:
I believe for the moment, the db changes will suffice and I will start writing some tests. These two models don't have a lot on them, so not a whole lot to validate at the moment. Maybe some minimum lengths for names and such would be advantageous and will address that when I start hammering out the tests.

matthewphilyaw committed:
adding tests for the new models

matthewphilyaw commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,149 @@
+  @missing_assoc_msg "does not exist"
+  @unique_msg "has already been taken"
+
+  defp do_setup do
```
There is a setup function that can be used that runs before each test, the reason I avoided it at the moment is because it's defined in the ModelCase and starts the db transaction which is imported via the using at the top.

I have few ideas around this that I need to review, but maybe we can clean thus up a bit.

copenhas replied:
ExUnit supports multiple setup functions. The [documentation](http://elixir-lang.org/docs/stable/ex_unit/ExUnit.Callbacks.html#content) kind of suggests the idea, but [looking at the source](https://github.com/elixir-lang/elixir/blob/v1.2.2/lib/ex_unit/lib/ex_unit/callbacks.ex#L108) for the macro looks like it does. Each time you call the `setup/2` macro it creates a uniquely named function and adds it to the setup list.

matthewphilyaw replied:
Awesome, I should be able to adjust my setup function and then take a context in each function instead.

matthewphilyaw committed:
removed model helper, drags coverage down and not being used.

matthewphilyaw commented:
@copenhas I think the suite of tests I have for now will suffice. May want to review and make sure I didn't miss something glaring.

In fact as I'm typing this the following come to mind
1. Character name could have spaces
2. Character name could be mixed case
3. Game name could have spaces
4. Game name could be mixed case

I'm not sure the case is an issue for either of these, but the spaces in the character name we may not want.

copenhas commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,149 @@
+  @game1   %{name: "abcdef", description: "abcdefghi"}
+  @game2   %{name: "ghijkl", description: "abcdefghi"}
+
+  @character %{name: "foobar", user_id: nil, game_id: nil}
```
Is there any significance to this one? Doesn't save you anything below and it's not named something like `valid_character` or `named_character`.

matthewphilyaw replied:
The name is a valid length and serves as the basis for all the tests that center around valid changesets and the combos where the game_id/user_id fire constraint errors. In those cases the name needs to be correct but the until you hit the DB you don't know if the foreign keys are correct for example.

Given that I refrained from calling it a valid_character, but perhaps there is better name - open for suggestions.

copenhas commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,149 @@
+    assert {:user, @missing_assoc_msg} in changeset.errors
+  end
+
+  test "Same character name can be used if the game_id is different with the same user_id" do
```
Your other tests started with a lowercase character, this one starts with an uppercase. Should pick a style. I believe our other tests start with lower.

matthewphilyaw replied:
Ah, yes good point. Pretty sure I copied and pasted below this point so the other picked it up from here. I'd have to review but I generally start with a lower case.

copenhas replied:
Also do you like the `it does this with that` style of writing tests? I would like to keep things consistent moving forward if we can. I believe we are using two different styles for test names.

matthewphilyaw replied:
Ah, yeah I seem to have the opposite style of `with this and this that happens` just get into that pattern when I'm writing cause it's how I'm thinking - parameters then outcome.

To be honest I don't mind rewriting to adapt which ever style and to stay consistent don't even realize I'm doing it so not sure that alone is good argument for me to continue nor am I'm sure I'm 100% consistent.

Maybe we can add a wiki page or something with style guides if you have some suggestions and maybe provide like just a few examples? I'm totally game for changing these if I need to. Rewording is not that big of a deal.

copenhas commented on test/support/model_case.ex:
```diff
@@ -30,30 +30,4 @@ defmodule Yggdrasil.ModelCase do
-  """
-  def errors_on(model, data) do
-    model.__struct__.changeset(model, data).errors
-  end
```
Why was this removed?

matthewphilyaw replied:
Primarily because it's not used and keeps getting reported on test coverage. That alone is not a good reason I suppose, but for example we use create_changeset on the User model, so calling this would also not work because it assumes `changeset`

and on the game/character stuff just didn't quite figure out how I wanted to use it. Can be resurrected of course but seems limiting with the fact that it assume "changeset" and we already have one case where we don't have that.

copenhas replied:
That's a good point about it assuming the function. Yeah just leave it out for now I guess.

copenhas commented on web/models/character.ex:
```diff
@@ -0,0 +1,29 @@
+    timestamps
+  end
+
+  @required_fields ~w(name user_id game_id)
```
I see these are required but I didn't see any tests documenting that. Also It looks like you can work with a `user` and `game` field which can contain the struct. Is there a reason all your code refers to the *_id fields and not the those?

matthewphilyaw replied:
Very good point, the only tests that hint to this are the ones around the assoc constraints and it just sort eludes to it so definitely need those cases in there prove they are required.

As far as why *_id fields, because the cast works with ids not the structs for now. I believe with ecto 2.0 - [this](https://github.com/elixir-lang/ecto/blob/master/CHANGELOG.md#improved-association-support) will enable what us to use the game/user fields instead of having to use ids.

from my testing for example this does not work assuming the required fields was adjusted to mark game/user instead of the ids like this `~w(name user game)`

``` elixir
test "valid character produces a valid changeset" do
    fixture = do_setup

    user = fixture.user
    game = fixture.game

    valid_character = %{@character | user: user, game: game}

    character = Character.changeset(%Character{}, valid_character)
    assert character.valid? == true
  end
```

That will fail because the cast will not find the association or something like that. There are also put_assoc functions etc, but they do not work with belongs_to for now anyway tried those as well. I think 2.0 will iron this out once it lands.

copenhas commented:
Oh, yeah we would want to allow mixed casing and spaces in the character and game names. Although probably would want the character's name to be case insensitive in the database so that the index wouldn't allow `Copenhas` and `copenHas`.

matthewphilyaw committed:
added two more cases for required fields and now have a propper setup function.

matthewphilyaw commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,150 @@
+  @unique_msg "has already been taken"
+  @required_msg "can't be blank"
+
+  setup _context do
```
@copenhas good call on the multiple setups. This was more inline with what I wanted.

matthewphilyaw commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,150 @@
+    assert {:user, @missing_assoc_msg} in changeset.errors
+  end
+
+  test "same character name can be used if the game_id is different with the same user_id", ctx do
```
Went with lower case here, as far as the style of the testing names word wise we may be better making a pass at all tests to standardize them than trying to do this piece mill.

matthewphilyaw commented on test/models/character_test.exs:
```diff
@@ -0,0 +1,150 @@
+    assert character.valid? == false
+  end
+
+  test "character changset requires game_id" do
```
@copenhas this test and the one below cover the required fields.

As I said in the previous comment, for now we are somewhat limited to using *_ids until ecto 2.0 when persisting chracters.

matthewphilyaw commented:
@copenhas sorry for rebase, needed master.

matthewphilyaw commented on test/channels/player_channel_test.exs:
```diff
@@ -1,62 +1,75 @@
   use Yggdrasil.ChannelCase
-  alias Yggdrasil.PlayerChannel
-  alias Yggdrasil.Message
+  alias Yggdrasil.{PlayerChannel, Message, User, Game, Character}
```
Elixir 1.2 introduced this syntax for importing aliases, there were enough here that I went for it, but maybe we can make a pass to switch to this new syntax. I think it's a little cleaner when you have several imports.

matthewphilyaw commented:
@copenhas ok so I switched the player_channel around to accept character_id instead of user id, so the client is expected to join a topic in the form of `player:123` where `123` is the character_id. 

I believe this alone will allow us to support multiple games, there is no need for game_id in the player registry key.

As far as when the player process starts and it interacts with the game - it has the character id, so it can find the game. A lot of details there we just don't have at the moment.

copenhas commented:
Cool. So it looks like you didn't have to change any of the player registry code.  You may want to still go in and change the variable names so they are consistent.

matthewphilyaw commented:
Ah, yes good point faintly remember some variables in there now that you mention it, I forgot to look in the player registry when I was doing this last night.

matthewphilyaw committed:
rename user_id to char_id in player modules

matthewphilyaw commented on lib/yggdrasil/player.ex:
```diff
@@ -25,19 +25,19 @@ defmodule Yggdrasil.Player do
         push_msg.(Message.info("Welcome to the game"))
         {:ok, %{
-          user: user_id,
+          user: char_id,
```
Hmmm, overlooked the "user" key here.

matthewphilyaw committed:
changer user key to character.

matthewphilyaw committed:
renamed mon2user functions to mon2character.

matthewphilyaw commented:
I believe I sorted out all the naming issues, a few slipped by grep because I wasn't look for just "user" when I did this originally but should be good now.

matthewphilyaw committed:
added game controller and view

matthewphilyaw committed:
super basic game controller tests to get started with

matthewphilyaw committed:
had one unused variable, and removed IEx.pry calls

matthewphilyaw commented on test/controllers/game_controller_test.exs:
```diff
@@ -0,0 +1,79 @@
+
+    db_games == api_games
+  end
+
```
@copenhas right now only have two tests that cover the happy case.

right now with only allowing get the only other major one we may want to cover is the case where there are no games, which should return an empty list to the client. I'm sure there are others as well,

matthewphilyaw committed:
character controller roughed out.

matthewphilyaw commented on web/controllers/character_controller.ex:
```diff
@@ -0,0 +1,60 @@
+  def index(conn, %{"filter" => %{"user_id" => user_id}}) do
+    # this should not happen, but could easily
+    # probably don't need to crash but rather return an error
+    ^user_id = conn.assigns.user
```
Pretty sure we need an error here instead of just crashing I think.

matthewphilyaw commented on web/controllers/character_controller.ex:
```diff
@@ -0,0 +1,60 @@
+
+    # returns nil if not found
+    # need to sort out what we want here.
+    char = Repo.one from c in Character,
```
Same kind of deal, this returns nil I think, not sure if we should error or just render an empty data field in the response...

matthewphilyaw commented on web/controllers/character_controller.ex:
```diff
@@ -0,0 +1,60 @@
+
+    # returns nil if not found
+    # need to sort out what we want here.
+    char = Repo.one from c in Character,
```
Obviously same concerns as in the show function.

matthewphilyaw commented on web/controllers/character_controller.ex:
```diff
@@ -0,0 +1,60 @@
+defmodule Yggdrasil.CharacterController do
```
Only thing I don't have yet is a patch handler for updates per JSONAPI's recommendation. Eventually will get there.

matthewphilyaw commented:
@copenhas may want to look this over a little. This may not be perfect yet, mainly trying to bang out the controllers that are "good enough" to give us characters and games and then come back to polish it at some point.

I need tests for character controller as well. Will try to work on that tomorrow.

matthewphilyaw commented on web/controllers/game_controller.ex:
```diff
@@ -0,0 +1,11 @@
+  alias Yggdrasil.Game
+
+  def index(conn, _params) do
+    games = Repo.all Game
```
@copenhas in-light of the changes you did with splitting the app, I don't think I want to do this in the controller.

I'm going to have to do some clean up and reorganize anyway with this change now in master so do you have maybe some suggestions on how we cleanly present these sorts of things to the web layer?

``` Exlir
# Almost in favor of having Game
Yggdrasil.Game.list_games()
Yggdrasil.Game.get_game(123)
Yggdrasil.Game.list_characters()
Yggdrasil.Game.get_character(123)
Yggdrasil.Game.add_character(...)
```

We do have a game module for the game struct, but either we can call that something like GameModel or use module names to differentiate (not sure how I feel about that yet, but may be an option) so like `Yggdrasil.Model.Game` for the game model, and `Yggdrasil.Game` for the api exposed to the web... Again GameModel maybe a better a choice there but suggesting both here.

copenhas replied:
What's the aversion to using the Ecto repo here? I'm not sure what the best practice should be so it might take some discussion.

The example you gave above is like making the `Game` module a repository doing nothing more but hiding your data access. I know with NHibernate that became an anti-pattern because it makes it more difficult to take full advantage of your data mapper and query library, in this case Ecto. The other issue is that you end up with a single module that have to keep changing as things need to retrieve the model differently as well as you have logic to tie to it. You can argue that it breaks SRP. I know in .NET land we might use query objects when something was reusable or to complex to stay inline. Then straight NHibernate usage otherwise. Experience has shown that only the most basic queries end up being reused, most seem to be specialized for a particular case. 

I think using Ecto directly here is fine, especially to start with. 

That doesn't handle the question of what's our convention for naming the data representation versus the process. I kind of like using `Yggdrasil.Models.*` for all the ecto schema stuff, or maybe call it `Yggdrasil.Schema.*`. Then you can organize the rest normally, but I think we'll want to be careful what we name things either way. `Yggdrasil.Schema.Room` and `Yggdrasil.Room` will be easy to confuse with an alias in the file. Maybe the data structure should win and be called the thing, then try to name the process something special? Maybe:

``` Elixir
Yggdrasil.Schema.Character
Yggdrasil.Schema.Room
Yggdrasil.ActiveRoom # or RoomServer
```

matthewphilyaw replied:
Ah no the goal isn't a repo but a game engine Api. This branch hasn't been updated yet but we split apps. 

I'd rather not let the game engine internsls bleed out of the Yggdrasil app so why I'm suggesting we introduce an api. 

So my goal is to let the game engine manage the life cycle of its core assets like character. I'm not sure the web needs to. 

So while I said Yggdrasil.Game the intent is a module that lets you interact with the game engine. 

So for example in my examples of chracater any modification or removing probably should be checked to ensure the character isn't in a running game or active and if so either error or notify the room etc. Also if there are any other things like creating a character internally updates several tables to place in a default room etc the game engine is best suited for this not the web. 

So add_character in my example above maybe a direct mapping temporarily it will likely expand. Thus we can do that and not modify the web. 

All that said internal to the game engine if you need a character im all for using a repo and totally agree that that is best. I just don't think we want to use repo directly outside of the app. Harder to test and spreads you game engine logic out beyond the game engine app. Also the web will have limited set of queries I think no need for full queryiny as it will only have very specific things it needs. I think. 

Naming wise I'm good with what you suggested.

copenhas replied:
Yeah but the game engine will never need to worry about removing characters and such, that'll be a web thing. Also the web will need abilities to search but the game engine will most likely only perform id based lookups. Other administration things like deactivating old accounts and such, the game engine doesn't care about any of that. 

The issue with a character requiring updating multiple tables might be handled for us when Ecto hits 2.0, but either way I get that we might need to write a module to hide that stuff behind. I do question if everything the web API will need should be in `Yggdrasil`. So maybe I should ask do you see `Yggdrasil` encompassing everything?

matthewphilyaw replied:
So I see Yggdrasil encompassing the game engine. The otp app to me isn't
just for code organization it symbolizes an independent running app.

I don't see it containing a user though. I think YggdrasilWeb should
contain the user and all administrative stuff. Possibly even having its own
repo.

So if we do this we achieve two isolated apps where the web interacts with
game engine through a module but doesn't know or care about its internals.

For example a character is something that I would say is a game engine
component. Lots of things in the game engine will depend on it and use it.
Thus to me it should own it and mange its life cycle.

To your point yes the web exposes an API to remove a character but I'll be
shocked if removing a character will not need some sort of logic wrapped
around it to ensure its not in an active room this being used if so either
error or notify the room. In addition there maybe other clean up and checks.

So if the web API handles this in the controller for example then the
controller has to know all this logic if it's going to directly manipulate
the repo. This also means tests for ensuring the stability on the running
game character are now in the controller. Because the game engine app can't
test it. It has no idea a character will be removed.

However if this is in an API like I suggested above Yggdrasil can ensure a
vet the management of a characters life cycle that it depends on. Makes it
easier to test I think. It keeps the test local to the pieces that depend
on it heavily.

It's my view anyway kind of how I have this split up. I think it's best to
think of otp apps as black boxes that expose its functionality.
On Thu, Feb 18, 2016 at 10:24 PM Sean Copenhaver notifications@github.com
wrote:

> In web/controllers/game_controller.ex
> https://github.com/voluspa/yggdrasil/pull/23#discussion_r53418983:
> 
> > @@ -0,0 +1,11 @@
> > +defmodule Yggdrasil.GameController do
> > -  use Yggdrasil.Web, :controller
> >   +
> > -  alias Yggdrasil.Game
> >   +
> > -  def index(conn, _params) do
> > -    games = Repo.all Game
> 
> Yeah but the game engine will never need to worry about removing
> characters and such, that'll be a web thing. Also the web will need
> abilities to search but the game engine will most likely only perform id
> based lookups. Other administration things like deactivating old accounts
> and such, the game engine doesn't care about any of that.
> 
> The issue with a character requiring updating multiple tables might be
> handled for us when Ecto hits 2.0, but either way I get that we might need
> to write a module to hide that stuff behind. I do question if everything
> the web API will need should be in Yggdrasil. So maybe I should ask do
> you see Yggdrasil encompassing everything?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/voluspa/yggdrasil/pull/23/files#r53418983.

matthewphilyaw replied:
Also it maybe easier for me to leave it as is finish this. Them once merged
to branch off and mock up what I'm thinking. It maybe easier to see and us
to discuss so I have something to point at.

That way if we say no then we still have this and can kill the other
branch.
On Thu, Feb 18, 2016 at 11:04 PM Matthew Philyaw matthew.philyaw@gmail.com
wrote:

> So I see Yggdrasil encompassing the game engine. The otp app to me isn't
> just for code organization it symbolizes an independent running app.
> 
> I don't see it containing a user though. I think YggdrasilWeb should
> contain the user and all administrative stuff. Possibly even having its own
> repo.
> 
> So if we do this we achieve two isolated apps where the web interacts with
> game engine through a module but doesn't know or care about its internals.
> 
> For example a character is something that I would say is a game engine
> component. Lots of things in the game engine will depend on it and use it.
> Thus to me it should own it and mange its life cycle.
> 
> To your point yes the web exposes an API to remove a character but I'll be
> shocked if removing a character will not need some sort of logic wrapped
> around it to ensure its not in an active room this being used if so either
> error or notify the room. In addition there maybe other clean up and checks.
> 
> So if the web API handles this in the controller for example then the
> controller has to know all this logic if it's going to directly manipulate
> the repo. This also means tests for ensuring the stability on the running
> game character are now in the controller. Because the game engine app can't
> test it. It has no idea a character will be removed.
> 
> However if this is in an API like I suggested above Yggdrasil can ensure a
> vet the management of a characters life cycle that it depends on. Makes it
> easier to test I think. It keeps the test local to the pieces that depend
> on it heavily.
> 
> It's my view anyway kind of how I have this split up. I think it's best to
> think of otp apps as black boxes that expose its functionality.
> On Thu, Feb 18, 2016 at 10:24 PM Sean Copenhaver notifications@github.com
> wrote:
> 
> > In web/controllers/game_controller.ex
> > https://github.com/voluspa/yggdrasil/pull/23#discussion_r53418983:
> > 
> > > @@ -0,0 +1,11 @@
> > > +defmodule Yggdrasil.GameController do
> > > -  use Yggdrasil.Web, :controller
> > >   +
> > > -  alias Yggdrasil.Game
> > >   +
> > > -  def index(conn, _params) do
> > > -    games = Repo.all Game
> > 
> > Yeah but the game engine will never need to worry about removing
> > characters and such, that'll be a web thing. Also the web will need
> > abilities to search but the game engine will most likely only perform id
> > based lookups. Other administration things like deactivating old accounts
> > and such, the game engine doesn't care about any of that.
> > 
> > The issue with a character requiring updating multiple tables might be
> > handled for us when Ecto hits 2.0, but either way I get that we might need
> > to write a module to hide that stuff behind. I do question if everything
> > the web API will need should be in Yggdrasil. So maybe I should ask do
> > you see Yggdrasil encompassing everything?
> > 
> > —
> > Reply to this email directly or view it on GitHub
> > https://github.com/voluspa/yggdrasil/pull/23/files#r53418983.

copenhas replied:
When I have setup a similar layer in the past it has ended up mostly being a pass through layer that replicated what the web needed to do. It didn't add a lot of a value and created a extra work.

I get that we may need to do a bunch of extra work on a character before we remove it, but it might be context specific. Example is that an admin comes in and removes a character, it should auto logout the player if they are playing, send them an email with the reason for removal, and delete the character. If a player came in and tried to delete themselves while they were player we most likely want to return back an error telling the player that they need to logout first as basically a means to ensure they didn't do it accidentally while trying to view something about their account while player. The game engine won't know these things and we could end up with a large API with little value.

I think it is simpler to start off without the layer of abstraction and move towards it if we start feeling the pain when making changes. Maintainable code is code that you can change easily, not just how things are organized. Sort of the YAGNI and evolutionary design approach. We have the luxury that there is not deadline and we can make these refactorings as we feel them.

I feel the same thing about splitting things out into separate apps too early. It adds concerns about ensuring we don't run into circular dependencies or confusing and leaky APIs because we made a bad assumption that something could be split out.

matthewphilyaw replied:
OK so this example of context specific to me is confusing the web with the game. That's what I'm trying to avoid.

So ok one I would never want APIS like this and this may be your concern

``` elixir
Yggdrasil.remove_logged_in_character()
Yggdrasil.remove_non_logged_in_character()
```

They are too specific. I don't think there is any value in taking something like remove_logged_in_character() and have it hide calls to check if a character is online and act differently if it is, or error differently. Totally agree the web knows best here. That is very context specific and indeed would be a pass through layer that would have to change all the time.

However on the flip side of that coin I certainly don't want a web controller knowing all the internals of a the game because for example how a character is determined to be online is something the game is best suited to answer, and how to shutdown a player process for that character is something that the game is best suited to answer and would know best.

Also how to remove a character from a game I think is best answered in the game app, it knows best how the game uses the character right? The game may require that in order to remove a character it can't be running, therefore it just simply doesn't allow it. The web can then say ok, I need to first shutdown the player process. 

These are things we will wrap up in modules anyway I believe and will be in Yggdrasil. 

So in your example with the context I full expect something like this

``` Elixir
# admin removing a character
if Yggdrasil.Player.online(character_id) do
  :ok = Yggdrasil.Player.shutdown(character_id)
end
Yggdrasil.Character.remove(character_id)

# User trying to do it
case Yggdrasil.Player.online(character_id) do
  true -> # notify user they can't
  false -> Yggdrasil.Character.remove(character_id)
end
```

Otherwise this code would balloon up with having to know how to figure out all the stuff involved with characters... I don't quite see why would would want that or the benefit to it. I may have the wrong perspective here. But for example if we didn't user Player module here the web would need to know it's an ETS and talk to the ETS table.

Also in my example there, there it's probably best that we don't have like one monolithic API, I'm more or less full expecting lots of modules from Yggdrasil to be used and the web can orchestrate it. In this example the same api calls are used just called differently. I'm aiming for small building blocks. 

I'm just not a fan of direct Repo access unless it's needed, because then the web has to know too much I think.

matthewphilyaw replied:
So for example

``` Elixir
# admin removing a character
if Yggdrasil.Player.online(character_id) do
  :ok = Yggdrasil.Player.shutdown(character_id)
end
Yggdrasil.Character.remove(character_id)

# User trying to do it
case Yggdrasil.Player.online(character_id) do
  true -> # notify user they can't
  false -> Yggdrasil.Character.remove(character_id)
end
```

Yggdrasil.Character.remove(character_id) can do checks to ensure that it's not in active player or running and error if so. It can be that simple in that it either succeeds or fails but doesn't do anything else, but protects the game. The web would be responsible for actually shutting down the player process etc.. 

However if we did this

``` Elixir
# admin removing a character
if Yggdrasil.Player.online(character_id) do
  :ok = Yggdrasil.Player.shutdown(character_id)
end
char = Yggdrasil.Repo.get Character, character_id
Yggdrasil.Repo.delete character

# User trying to do it
case Yggdrasil.Player.online(character_id) do
  true -> # notify user they can't
  false -> 
    char = Yggdrasil.Repo.get Character, character_id
    Yggdrasil.Repo.delete character
end
```

So two code paths may have direct access to remove character, and we have to rely that they will both check the active players and more testing rather duplication of tests in my opinion.

Don't know I just was thinking that if we did what the first form we could then test out Yggdrasil.Character.remove() to ensure it only succeeds if the constraints are met and for example it's not actively being used. This way we can ensure one code path and consistency.

Anyway to your point it's not like the web will not have to know how to use the game engine, it's going to have to know how to orchestrate it etc and I don't think there is any value in trying to hide that at all. In this example even the web will have to know to check the player is online, and then shutdown/logout the player and also remove the character. Wrapping that up into single function call is very bad I think, because just as you pointed out the set of events may differ from context.

copenhas replied:
OK, I think I'm in pretty much complete agreement with your examples. Exactly, you can't push too much behind the API wall or you'll end up with overly specific functions. So the question comes down to direct repo access. It offers any other sub-systems complete flexibility in how to retrieve data, but to your point we could end up with messy code.

Perhaps as convention we allow queries to the repo only at first? Other APIs are driven off of the struct returned and our module APIs. It could also be convention that you can use the repo directly until it's not just straight access, then logic needs to be pushed down. So if all you really needed to do was straight CRUD then fine, but after that you gotta push the logic down into the modules. Hell the repo is public anyway and usage is only effecting us.

matthewphilyaw replied:
Yeah, more of a fan of a suite of functions that do one thing well that can be composed than overly specific functions where it's like oh you need to call this one for that action or this other for that action and they may be disjoint where you need something in between.

A user for example I would rather the web completely own and have direct repo access, not that I'm shy or subtle about how I want to throw that back into the web :wink:.

I do have hard time thinking of any one example where you would be doing straight CUD (doesn't read very well without the R...haha) on any game related table though that the game isn't better suited in doing.

My opposition to direct access to the Game's repo is that it allows unchecked abilities to remove or update records that the web will most likely never know if it can safely do. That said I do see where the query abilities would be needed definitely.

That said, it's not the repo that offers the querying, it's the Ecto.Query module, so there is no reason the APIs can't take in a ecto query to execute and return results. I'm not opposed to exposing the ecto schemas for such things or use else where. 

Even still I have hard time picturing any game related tables that will be as simple as getting only data from the tables. Like if you want an admin to see characters in a game, but you want to allow the admin to say show me only characters actively playing the game. That is no longer just a db question, and maybe the it's a Character.query() function at that point where you give ecto query and it augments the results coming back to include an active flag which the UI can use to filter for example or can filter before the UI.

I was sort taking the stance that it's easier to slowly open up and expose access to underlying things than start wide open and trying to reign it back in. Like any case where what we provide doesn't satisfy a certain need we can step back and say ok what combo is needed what is the impact of that, and we can test it out isolated instead of through controllers etc. Then we end up with the bare minimum api we need as we go.

All that said, I'm ok if you prefer to start with direct repo access and iteratively pull it back into yggdrasil as well. I see your points and concerns as well.

matthewphilyaw replied:
Btw I'm also not in favor of this any more

``` Elixir
Yggdrasil.Game.list_games()
Yggdrasil.Game.get_game(123)
Yggdrasil.Game.list_characters()
Yggdrasil.Game.get_character(123)
Yggdrasil.Game.add_character(...)
```

What ever this turns into, I think it's ok to have several modules that are composed instead of one like this. If we determine a need to put a process around character creation for example `Yggdrasil.Character.create()` or something along those lines but point being that I think having task specific modules are probably better than one module where the world of api calls hangs off of.

copenhas replied:
I'm thinking at first a lot of things will probably be straight CRUD and eventually several things will be more and more complex. That's probably why I favor starting with direct repo calls. 

I do agree the big thing is that we'll want full querying abilities, so the least we that exposed. It's possible that there are other things.

I know with NHibernate a big reason you want to push the session access up in the layers is because you can take advantage of batched queries, eager/lazy loading, but I can't think of anything around CUD that benefits from direct access since the models and the mapping take care of that (which Ecto 2.0 will improve). 

Anyway, I'm cool with your idea of wrapping up the CRUD with the module but allowing the "Read" function to take an ecto query. You addressed my primary concerns in previous comments.

I do want to avoid using primitives everywhere though. Maybe our modules should work off of the structs, so you first gotta query what you want.

matthewphilyaw replied:
Good point about eager/lazy loading, I actually think that can be entirely handled in a query according to [this](https://hexdocs.pm/ecto/Ecto.Query.html#preload/3). If you scroll down a bit it mentioned the use directly in a query. 

Ecto seems to be split up nicely, so like for example Sessions in Nhibernate provide the tracking of changes and batching of those changes, where as Ecto relies on changesets and I don't think the Repo keeps track of any of that. Same with querying, the Repo doesn't provide it just executes the query passed in.

So it will be interesting to see how we organize compared to our experiences with Nhibernate. I'm really curious to see and see what they bring in 2.0.

I also completely agree with the use of primitives, that should be restricted to only places where it's needed which hopefully is rare.

copenhas replied:
:+1:

matthewphilyaw committed:
update dir structure.

matthewphilyaw committed:
removed ecto.query

matthewphilyaw committed:
controllers updated to match app structure.

matthewphilyaw commented:
@copenhas rebased on master to pull in the new dir structure. 

Going to try to knock out that last test case for character controller. For now I left the direct repo access in the controllers. Ideally would like to work on taking that out at some point, but it was already there so and would rather get this merged in to give us characters and games.

matthewphilyaw commented:
Also coverage drop is due to character controller that has no tests, which I'm working on.

matthewphilyaw committed:
added the start of character controller tests.

matthewphilyaw committed:
bump guardian version

matthewphilyaw committed:
Load user via Guardian.Plug.LoadResource on each request

This found some issues with the way I defined the YggdrasilWeb.GuardianSerializer.

I have made appropriate changes and modified tests to reflect them.

matthewphilyaw committed:
add controller tests, which also needed a character_view module

matthewphilyaw commented:
@copenhas  you may want to read over this if you get time. I will put some more comments in later tonight there are somethings that I'm sure you will have questions about and want to bring up.

Anyway I have added two tests so far one for listing out all the characters for a user and one for getting one character by id. When I get some time tonight I will explain the route I went.

Still need to cover create/delete. I don't have update at the moment but will add that as we need it.

There are also some  tests around ensuring that the characters coming back are for the user that is logged in and not just any user. So lots of good stuff left to test but it's a start and also already has shaken out a few things around our auth process.

matthewphilyaw commented on apps/yggdrasil_web/test/controllers/character_controller_test.exs:
```diff
@@ -0,0 +1,96 @@
+  test "get /api/characters/char_id returns the correct character", ctx do
+    char = Enum.at ctx.chars, 0
+
+    char_json = CharacterView.render("show.json", conn: %{}, data: char)
```
After studying how JaSerializer tests their stuff noticed that they just use the view directly like this. 

That said I think JaSerializer has format method that it uses internally that accepts a connection and formater. 

Seeing this example showing that connection can be empty I think we could use it directly and pass the view in as the formatter. Not sure what is best here but this works better than hand coding a solution like I have in other tests. 

I think after this branch I need to make a sweep at our tests and make them more consistent.

matthewphilyaw commented on apps/yggdrasil_web/web/router.ex:
```diff
@@ -10,6 +10,7 @@ defmodule YggdrasilWeb.Router do
   pipeline :guardian do
     plug Guardian.Plug.VerifyHeader
     plug Guardian.Plug.EnsureAuthenticated, handler: YggdrasilWeb.GuardianErrorHandler
+    plug Guardian.Plug.LoadResource
```
This whole block with the addition of LoadResource I think replaces the need for [this line](https://github.com/voluspa/yggdrasil/blob/master/apps/yggdrasil_web/web/controllers/session_controller.ex#L16) in the session controller. 

That line called api_signin but since each request through this pipeline digests the auth header and effectively does the same thing I don't believe the sign in is needed anymore as after login the first request that needs auth with effectively sign the user on via this block and make the user available down stream of his call just as the sign in would. 

This also means that if you have a valid token you don't have to explicitly call sign in. I'm ok with this. Since tokens have an expiration and checked in each time anyway and we don't have a real session. 

Any objections? This means the user could be idle and their connection be cleaned up and later on within that tokens expiration call the API without logging in explicitly via the login API call. For example the user comes back and refreshes their list of characters the next day it would just work not ask them to login provided the token is valid.

copenhas commented on apps/yggdrasil/test/yggdrasil/character_test.exs:
```diff
@@ -0,0 +1,156 @@
+defmodule Yggdrasil.CharacterTest do
```
We need to figure out a way to organize our tests more. We'll end up with several functions in a module and need to have an easy way to group them. Is it possible to have sub-modules with tests?

``` Elixir
defmodule Yggdrasil.CharacterTest do
    use ExUnit.Case, async: false

    defmodule Yggdrasil.CharacterTests.CreateChangesetTest do
        use ExUnit.Case, async: false

        test "it requires a valid user id" do
        end
    end
end
```

I'm just thinking that as we add functions this file will get ugly. Also I greatly prefer the BDD spec style test names. So I favor "it requires the character name to be at least 3 characters" as a test name over "invalid character name length < 3".

matthewphilyaw replied:
So I'm not a fan of the submodules in one file anyway and I believe we tried this in the bowling kata and it didn't work.

Just peeked at the plug test suite for the Conn struct cause it's massive and sort of case like you just mentioned and a lot of tests are like "foobar/3 is something" so huge file but seems how it's organized.

So makes me wonder if we maybe do something of a hybrid where we have a set of macros that expands the tests for us.

we write something like 

``` elixir
for "Caracter.changeset/2" do
  it "requires a valid user id" do

  end
  it "requires a valid game_id" do

  end
end

# which expands when run to 
test "Character.changeset/2 requires a valid user id" do

end

test "Character.changeset/2 requires a valid game_id" do

end
```

Maybe we can augment the printing as well haven't dug that deep. There maybe better language but I'm ok with each test when they fail to print out the "Character.changeset/2" each time, because it's just easier to find after several failures in one file than having to scroll back up in the console.

I wonder if that would be that hard to do? Any interest? I do agree we need organization.

Then we can obviously do a directory for the module with submodules in it in separate files if gets really big.

matthewphilyaw replied:
See your note with describe so with describe the macro could be 

``` Elixir
describe "Caracter.changeset/2" do
  it "requires a valid user id" do

  end
  it "requires a valid game id" do

  end
end
```

That said when it's printed out I don't really care, practically having it repeat like this

```
"Caracter.changeset/2 requires a valid user id"
"Caracter.changeset/2 requires a valid game id"
```

makes it easier to understand the entire test you are looking without having to scroll up to the top and see what "it" is, but we may be able to change the test output as well.

I did omit the describe character line, don't think it's needed as it would be the module name which will be printed out and that would mean the entire test would be nested to simply have the describe at the top.

copenhas commented on apps/yggdrasil/test/yggdrasil/character_test.exs:
```diff
@@ -0,0 +1,156 @@
+    {:ok, _character} = Repo.insert character
+  end
+
+  test "invalid character name length < #{@min_len}" do
```
I didn't see a test for a max length. We should restrict the lengths of the names (character and game). There could be little issues besides storage to think about such as formatting names on the screen, or super long names flooding the console.

matthewphilyaw replied:
Yeah, true. Didn't enforce one, thus no test.

Can certainly enforce it on the changeset, not sure there is need to have postgres do this - from what I have read it seems to be very efficient with variable length columns so I believe ours are string? Unless we just need a hard stop there.

I'd imagine maybe 50 characters or less is probably enough don't know need to google that I guess.

copenhas commented on apps/yggdrasil/test/yggdrasil/character_test.exs:
```diff
@@ -0,0 +1,156 @@
+    assert {:user, @missing_assoc_msg} in changeset.errors
+  end
+
+  test "same character name can be used if the game_id is different with the same user_id", ctx do
```
As another example of test names: "it requires the character name to be unique for a game" and "it does not require character names to be unique across games".

If the test names were printed to screen it should be pretty readable. As an example for a full BDD spec printed out:

```
    describe Character
        describe #changeset/2
            it requires character name to be at least 3 characters
            it requires a valid user
            it requires a valid game
            it does not require character name to be unique across games
            it requires character name to be unique inside a game
```

matthewphilyaw committed:
added coverage for create/delete and ensured the others tests fulfilled what was stated.

matthewphilyaw commented on apps/yggdrasil_web/test/controllers/character_controller_test.exs:
```diff
@@ -0,0 +1,147 @@
+  test "post /characters creates a character for a given user", ctx do
+    user = Enum.at ctx.users, 0
+    game = Enum.at ctx.games, 0
+    char = %{"name" => "unique_tester", "game-id" => game.id} # server fills in user_id
```
Server knows what user is posting this, so it assumes the user_id needs to be that and just clobbers what the client sent and sets it on the server. I also need another test to prove this out btw.

The obvious concern here is an admin wanted to create a character for a user he can't. The counter to that, is that we don't want a user creating characters for other users.

The larger question here, which also applies to the other actions in this /characters endpoint in general is should we have two? I'll point this out in the character file may be more clear there.

matthewphilyaw commented on apps/yggdrasil_web/web/controllers/character_controller.ex:
```diff
@@ -0,0 +1,58 @@
+  alias Yggdrasil.Character
+
+  def index(conn, _params, user, _claims) do
+    user_id = user.id
```
@copenhas Ok, so with this controller I had a to make a decision either assume the user in the token was the driving this or let the api provide the user_id. 

If the api provides the api, then I obviously need to verify the user_id against the token to ensure that the user is requesting just there stuff.

The corner case here is the Admin, who may either want to see just there characters or all characters. So this would basically require permissions at that point.

With permissions, the api can then make a decision on what to return based on it.

However there are subtle differences here, like creating a user for example would require a user_id to create it with, and for a normal user this would be their id and should never be anyone else's, for and admin could be either. Same for deletes, the character to be deleted for a normal user needs to be the user id in that toke not anyone else's.

So my question is would this case be special enough that it would be simpler possibly more secure to have two routes.

```
/characters
/admin/characters
```

The first one is what we have now and basically all operations the server fills in the user_id on the server based on the token. One this eliminates the checks to ensure any passed in value matches the token, and two simplifies things.

The second one when we get to that point could offer a more expansive set of things you can do that we would not normally want. In which case we can one guard that route with permissions, and two once entered we know that this route is special so the entire controller can make the assumption that this is an admin vs one controller handling both.

Thoughts? The current implementation assumes that all actions on the controller are for one user, and is scoped to the token's user_id. It was the safer of the two approaches without a lot of extra code and me being a littler conservative at the moment.

copenhas replied:
Would it be better to add a permission system and have the API accept the user id? Seems like a better API would be to include the user id in the url anyway, isn't that more conventional?

```
/characters # all characters I can see
/characters?userid=12 # characters only for user 12
```

matthewphilyaw replied:
I would tend to agree with your examples.

Initially my concern was muddying up the code in the controller trying to determine what was intended but you can pattern match out the query params so may not be too bad honestly.

```
/characters # all characters I can see
#json api filter format but still only showing all for user 12
/characters?filter[user_id]=12 
```

So my only concern is that the calls `/character` and `/character?filter[user_id]=12` produce the same result for user 12 if user 12 can only see his characters, which leads to my question.

Should we possibly restrict `/characters` to a permission that can list all characters such that a user that isn't an admin for example can't see anything at all? This way we enforce that if you want to see just yours you 'must' call `/characters?filter[user_id]=12`. It feels like it may be slightly more consistent this way enforcing that idea.

Any thoughts on that? Just something I noted, it seems to lead to some ambiguity because a non admin user sees the same users either way and either call is valid for the purposes of seeing only the users characters and with and admin user only the one with the filter is valid for seeing only that users characters.

May not matter at all, just though it maybe worth mentioning.

copenhas replied:
I would rather have the same API and enforce everything via permissions then to have separate APIs. Also I don't see the issue with `/characters` and `/characters?filter[user_id]=12` returning the same results. I would think in most cases things would be readable to those signed in anyway.

matthewphilyaw replied:
Yeah, I think this is where documentation is going to be vital.

For example `/characters?filter[user_id]=...` should be used when wanting the specifics user's characters and not rely `/characters` and the users permissions. But this is entirely on the client and the more I think about it I think it's not really an issue. Vilive for example would probably use the filtered form if has a page to manage the user's character and then can show more by removing the filter, it's apparent now that it really is the client's responsibility.

So yeah I agree with you, one API controlled with permissions.

So do we want to do that in this round or accept this in a slightly broken state to collapse this down and get characters/games into master and then tackle permissions. Permissions are going to bring up a lot of questions I think. Purely looking at some guidance here, this PR is getting lengthy.

matthewphilyaw committed:
more tests to ensure a character can only be created for a given user and that a character can only be removed if it belongs to a given user. A given user being the one specified in the token.

matthewphilyaw committed:
Char.perms

Adds roles and permissions. This includes facilities to check that permissions are granted and a plug to control access to controllers based on permissions.
